File: gallery_wall_planner/utils/even_spacing.py
import tkinter as tk
from tkinter import ttk, messagebox
from typing import List, Dict, Tuple, Optional
from gallery_wall_planner.gui.wall_canvas import WallCanvas
from gallery_wall_planner.models.artwork import Artwork

def apply_even_spacing(wall_canvas: WallCanvas, imported_artworks: List[Artwork]) -> None:
    """
    Apply even spacing to selected artworks on the wall canvas with interactive UI.

    Features:
    - Visual selection of artworks with order tracking
    - Real-time calculation of total width
    - Customizable X and Y positions
    - Input validation

    Args:
        wall_canvas: The WallCanvas instance where artworks are displayed
        imported_artworks: List of imported artworks to be spaced
    """

    # Constants
    DEFAULT_Y_POSITION = 62  # Default height in inches
    MIN_SPACING = 0.5  # Minimum spacing between artworks in inches

    class SpacingUI:
        def __init__(self, master):
            self.master = master
            self.selected_artworks: List[Artwork] = []
            self.selection_order: Dict[int, int] = {}
            self.setup_ui()

        def setup_ui(self):
            """Initialize all UI components."""
            self.master.title("Even Spacing Tool")
            self.master.geometry("500x650")  # Increased height for Y position input
            self.master.attributes("-topmost", True)

            # Main container
            main_frame = ttk.Frame(self.master)
            main_frame.pack(fill="both", expand=True, padx=10, pady=10)

            # Input section
            self.setup_input_section(main_frame)

            # Position section
            self.setup_position_section(main_frame)

            # Artwork selection
            self.setup_selection_section(main_frame)

            # Info display
            self.setup_info_section(main_frame)

            # Action buttons
            self.setup_action_buttons(main_frame)

        def setup_input_section(self, parent):
            """Setup spacing boundary inputs."""
            input_frame = ttk.LabelFrame(parent, text="Spacing Boundaries (inches)")
            input_frame.pack(fill="x", pady=(0, 10))

            # Left boundary
            ttk.Label(input_frame, text="Left Boundary:").grid(row=0, column=0, padx=5, pady=5, sticky="w")
            self.left_entry = ttk.Entry(input_frame)
            self.left_entry.grid(row=0, column=1, padx=5, pady=5, sticky="ew")
            self.left_entry.insert(0, "0")

            # Right boundary
            ttk.Label(input_frame, text="Right Boundary:").grid(row=1, column=0, padx=5, pady=5, sticky="w")
            self.right_entry = ttk.Entry(input_frame)
            self.right_entry.grid(row=1, column=1, padx=5, pady=5, sticky="ew")
            self.right_entry.insert(0, f"{wall_canvas.wall.width:.1f}")

            input_frame.columnconfigure(1, weight=1)

        def setup_position_section(self, parent):
            """Setup position inputs including new Y position field."""
            position_frame = ttk.LabelFrame(parent, text="Artwork Position")
            position_frame.pack(fill="x", pady=(0, 10))

            # Y Position
            ttk.Label(position_frame, text="Center height from floor (standard: 62\")", font=("TkDefaultFont", 8)).grid(row=1, column=1, sticky="w", padx=5)
            self.y_entry = ttk.Entry(position_frame)
            self.y_entry.grid(row=0, column=1, padx=5, pady=5, sticky="ew")
            self.y_entry.insert(0, f"{DEFAULT_Y_POSITION}")  # Default value

            # Add tooltip/info label
            ttk.Label(position_frame,
                     text="Distance from bottom of wall",
                     font=("TkDefaultFont", 8)).grid(row=1, column=1, sticky="w", padx=5)

            position_frame.columnconfigure(1, weight=1)

        def setup_selection_section(self, parent):
            """Setup artwork selection listbox."""
            selection_frame = ttk.LabelFrame(parent, text="Select Artworks (click to select order)")
            selection_frame.pack(fill="both", expand=True, pady=(0, 10))

            # Listbox with scrollbar
            self.listbox = tk.Listbox(
                selection_frame,
                selectmode="multiple",
                height=10,
                activestyle="none",
                font=("TkDefaultFont", 10)
            )

            scrollbar = ttk.Scrollbar(selection_frame, orient="vertical", command=self.listbox.yview)
            self.listbox.config(yscrollcommand=scrollbar.set)

            scrollbar.pack(side="right", fill="y")
            self.listbox.pack(side="left", fill="both", expand=True)

            # Populate listbox
            for artwork in imported_artworks:
                self.listbox.insert(tk.END, f"{artwork.name} ({artwork.width}\" × {artwork.height}\")")

            self.listbox.bind("<<ListboxSelect>>", self.on_artwork_select)

        def setup_info_section(self, parent):
            """Setup information display labels."""
            info_frame = ttk.Frame(parent)
            info_frame.pack(fill="x", pady=(0, 10))

            self.selected_label = ttk.Label(info_frame, text="Selected: 0")
            self.selected_label.pack(side="left", padx=5)

            self.width_label = ttk.Label(info_frame, text="Total Width: 0.0\"")
            self.width_label.pack(side="left", padx=5)

            self.spacing_label = ttk.Label(info_frame, text="Spacing: -")
            self.spacing_label.pack(side="right", padx=5)

        def setup_action_buttons(self, parent):
            """Setup action buttons."""
            button_frame = ttk.Frame(parent)
            button_frame.pack(fill="x")

            ttk.Button(
                button_frame,
                text="Apply Spacing",
                command=self.apply_spacing,
                style="Accent.TButton"
            ).pack(side="left", padx=5)

            ttk.Button(
                button_frame,
                text="Cancel",
                command=self.master.destroy
            ).pack(side="right", padx=5)

        def get_y_position(self) -> Optional[float]:
            """Validate and return Y position input."""
            try:
                y_pos = float(self.y_entry.get())
                if y_pos < 0 or y_pos > wall_canvas.wall.height:
                    messagebox.showerror("Error",
                                       f"Y position must be between 0 and {wall_canvas.wall.height} inches")
                    return None
                return y_pos
            except ValueError:
                messagebox.showerror("Error", "Please enter a valid number for Y position")
                return None

        def on_artwork_select(self, event):
            """Handle artwork selection with order tracking that toggles and renumbers."""
            selected_idx = self.listbox.curselection()

            if not selected_idx:  # If nothing is selected, do nothing
                return

            clicked_idx = selected_idx[-1]  # Get the most recently clicked item

            # Toggle selection - if already selected, remove it
            if clicked_idx in self.selection_order:
                # Remove the clicked item and renumber remaining items
                removed_order = self.selection_order.pop(clicked_idx)
                # Decrement orders higher than the removed one
                self.selection_order = {
                    idx: (order if order < removed_order else order - 1)
                    for idx, order in self.selection_order.items()
                }
            else:
                # Add the clicked item with the next available order number
                next_order = len(self.selection_order) + 1
                self.selection_order[clicked_idx] = next_order

            # Update the selected artworks list based on current selection order
            sorted_selection = sorted(self.selection_order.items(), key=lambda x: x[1])
            self.selected_artworks = [imported_artworks[idx] for idx, _ in sorted_selection]

            # Update the display to show current numbering
            self.update_display()

        def update_display(self):
            """Update all display elements."""
            # Update listbox with order indicators
            self.listbox.delete(0, tk.END)
            for idx, artwork in enumerate(imported_artworks):
                order = f" {self.selection_order[idx]}" if idx in self.selection_order else ""
                self.listbox.insert(tk.END, f"{artwork.name} ({artwork.width}\" × {artwork.height}\"){order}")

            # Update info labels
            total_width = sum(art.width for art in self.selected_artworks)
            self.selected_label.config(text=f"Selected: {len(self.selected_artworks)}")
            self.width_label.config(text=f"Total Width: {total_width:.1f}\"")

            # Calculate and display projected spacing
            try:
                left = float(self.left_entry.get())
                right = float(self.right_entry.get())
                if right > left and self.selected_artworks:
                    spacing = (right - left - total_width) / (len(self.selected_artworks) + 1)
                    self.spacing_label.config(text=f"Spacing: {spacing:.1f}\"")
            except ValueError:
                pass

        def validate_inputs(self) -> Tuple[Optional[float], Optional[float]]:
            """Validate and return spacing boundaries."""
            try:
                left = float(self.left_entry.get())
                right = float(self.right_entry.get())

                if left < 0 or right > wall_canvas.wall.width:
                    messagebox.showerror("Error", "Boundaries must be within wall dimensions.")
                    return None, None

                if left >= right:
                    messagebox.showerror("Error", "Left boundary must be less than right boundary.")
                    return None, None

                return left, right
            except ValueError:
                messagebox.showerror("Error", "Please enter valid numbers for boundaries.")
                return None, None

        def apply_spacing(self):
            """Apply the calculated spacing to selected artworks."""
            left, right = self.validate_inputs()
            if left is None:
                return

            y_pos_center = self.get_y_position()
            if y_pos_center is None:
                return

            if not self.selected_artworks:
                messagebox.showerror("Error", "Please select at least one artwork.")
                return

            total_width = sum(art.width for art in self.selected_artworks)
            available_space = right - left
            spacing = (available_space - total_width) / (len(self.selected_artworks) + 1)

            if spacing < MIN_SPACING:
                if not messagebox.askyesno(
                    "Warning",
                    f"Calculated spacing ({spacing:.1f}\") is very small. Continue anyway?"
                ):
                    return

            # Position artworks
            current_x = left + spacing

            for artwork in self.selected_artworks:
                # Calculate position in inches (wall coordinates)
                artwork.x = current_x
                # Calculate Y position so center is at y_pos_center
                artwork.y = y_pos_center - (artwork.height / 2)

                # Debug output
                # center_x = artwork.x + (artwork.width / 2)
                # center_y = artwork.y + (artwork.height / 2)
                # print(f"Positioning {artwork.name}:")
                # print(f"  Center X (inches): {center_x:.1f}")
                # print(f"  Center Y (inches): {center_y:.1f} (from bottom: {y_pos_center:.1f})")
                # print(f"  Artwork size: {artwork.width}\" × {artwork.height}\"")

                # Check boundaries - now checking if entire artwork fits
                artwork.x, artwork.y = wall_canvas.enforce_boundaries_even_spacing(
                    artwork.x, artwork.y, 
                    artwork.width, artwork.height
                )
                
                # print(f"  After boundary check: X={artwork.x:.1f}, Y={artwork.y:.1f}")

                # Check if artwork already exists on canvas
                if artwork.id in wall_canvas.draggable_items:
                    # Update existing draggable
                    draggable = wall_canvas.draggable_items[artwork.id]
                    draggable.wall_object.x = artwork.x
                    draggable.wall_object.y = artwork.y
                    draggable.update_position()
                else:
                    # Add new draggable
                    wall_canvas.add_draggable(artwork)

                current_x += artwork.width + spacing
            
            self.master.destroy()

    # Create and run the UI
    root = tk.Toplevel()
    SpacingUI(root)
    root.mainloop()

File: static/js/even_spacing.js
// static/js/even_spacing.js
export class EvenSpacing {
    constructor(editor) {
        this.editor = editor;
        this.DEFAULT_Y_POSITION = 62; // Default height in inches
        this.MIN_SPACING = 0.5; // Minimum spacing between artworks in inches
        this.selectedArtworks = [];
    }

    init() {
        // Remove old modal if it exists
        const oldModal = document.getElementById('evenSpacingModal');
        if (oldModal) {
            oldModal.parentNode.removeChild(oldModal);
        }
        this.setupUI();
        this.bindEvents();
    }

    setupUI() {
        this.modal = document.createElement('div');
        this.modal.className = 'modal fade';
        this.modal.id = 'evenSpacingModal';
        this.modal.innerHTML = `
            <div class="modal-dialog modal-lg">
                <div class="modal-content">
                    <div class="modal-header py-2">
                        <h5 class="modal-title">Even Spacing</h5>
                        <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
                    </div>
                    <div class="modal-body" style="max-height: 70vh; overflow-y: auto;">
                        <div class="row g-3">
                            <!-- Left: Spacing controls -->
                            <div class="col-md-6">
                                <label class="form-label mb-1">Spacing Boundaries (inches)</label>
                                <div class="input-group input-group-sm mb-1">
                                    <span class="input-group-text">Left:</span>
                                    <input type="number" class="form-control" id="leftBoundary" value="0" step="0.1">
                                </div>
                                <div class="input-group input-group-sm mb-2">
                                    <span class="input-group-text">Right:</span>
                                    <input type="number" class="form-control" id="rightBoundary" value="${this.editor.wall.width}" step="0.1">
                                </div>
                                <label class="form-label mb-1">Center height from floor</label>
                                <div class="input-group input-group-sm mb-1">
                                    <input type="number" class="form-control" id="yPosition" value="${this.DEFAULT_Y_POSITION}" step="0.1">
                                    <span class="input-group-text">inches</span>
                                </div>
                                <div class="form-text small mb-1">Standard: 62" from floor</div>
                                <div class="d-flex justify-content-between small mb-1">
                                    <span id="selectedCount">Selected: 0</span>
                                    <span id="totalWidth">Total Width: 0.0"</span>
                                    <span id="spacingInfo">Spacing: -</span>
                                </div>
                            </div>
                            <!-- Right: Select Artworks -->
                            <div class="col-md-6">
                                <label class="form-label mb-1">Select Artworks (click to select order left to right)</label>
                                <select multiple class="form-select form-select-sm" id="artworkList" size="14" style="width: 100%;">
                                    ${this.editor.placedArtworks.map(artwork => 
                                        `<option value="${artwork.id}">${artwork.name} (${artwork.width}" × ${artwork.height}")</option>`
                                    ).join('')}
                                </select>
                            </div>
                        </div>
                    </div>
                    <div class="modal-footer py-2">
                        <button type="button" class="btn btn-sm btn-secondary" data-bs-dismiss="modal">Cancel</button>
                        <button type="button" class="btn btn-sm btn-primary" id="applySpacing">Apply Spacing</button>
                    </div>
                </div>
            </div>
        `;
        
        document.body.appendChild(this.modal);
        this.modalInstance = new bootstrap.Modal(this.modal);
    }

    bindEvents() {
        const artworkList = this.modal.querySelector('#artworkList');
        
        // Artwork selection - handle Ctrl/Cmd+click
        artworkList.addEventListener('click', (e) => {
            if (e.target.tagName === 'OPTION') {
                e.preventDefault();
                const option = e.target;
                const artworkId = parseInt(option.value);
                
                // Find if this artwork is already selected
                const existingIndex = this.selectedArtworks.findIndex(a => a.id === artworkId);
                
                if (existingIndex >= 0) {
                    // Remove from selection
                    this.selectedArtworks.splice(existingIndex, 1);
                    option.selected = false;
                } else {
                    // Add to selection (at end)
                    const artwork = this.editor.placedArtworks.find(a => a.id == artworkId);
                    if (artwork) {
                        this.selectedArtworks.push(artwork);
                        option.selected = true;
                    }
                }
                
                this.updateDisplay();
            }
        });

        // Boundary inputs
        ['leftBoundary', 'rightBoundary'].forEach(id => {
            this.modal.querySelector(`#${id}`).addEventListener('input', () => {
                this.updateDisplay();
            });
        });

        // Apply button
        this.modal.querySelector('#applySpacing').addEventListener('click', () => {
            this.applySpacing();
        });
    }

    updateSelection() {
        const select = this.modal.querySelector('#artworkList');
        const selectedOptions = Array.from(select.selectedOptions);
        
        // Get selected artworks in the order they were selected
        this.selectedArtworks = selectedOptions.map(option => {
            return this.editor.placedArtworks.find(a => a.id == option.value);
        });
    }

    // updateDisplay() method to show selection order
    updateDisplay() {
        const select = this.modal.querySelector('#artworkList');
        const options = Array.from(select.options);
        
        // Update all options to show selection order
        options.forEach(option => {
            const artworkId = parseInt(option.value);
            const index = this.selectedArtworks.findIndex(a => a.id === artworkId);
            
            // Remove any existing order number
            let text = option.text.replace(/ \(\d+\)$/, '');
            
            // Add order number if selected
            if (index >= 0) {
                text += ` (${index + 1})`;
            }
            
            option.text = text;
            option.selected = index >= 0;
        });

        // Update info labels
        const totalWidth = this.selectedArtworks.reduce((sum, artwork) => sum + artwork.width, 0);
        this.modal.querySelector('#selectedCount').textContent = `Selected: ${this.selectedArtworks.length}`;
        this.modal.querySelector('#totalWidth').textContent = `Total Width: ${totalWidth.toFixed(1)}"`;

        // Calculate and display projected spacing
        try {
            const left = parseFloat(this.modal.querySelector('#leftBoundary').value);
            const right = parseFloat(this.modal.querySelector('#rightBoundary').value);
            
            if (right > left && this.selectedArtworks.length > 0) {
                const spacing = (right - left - totalWidth) / (this.selectedArtworks.length + 1);
                this.modal.querySelector('#spacingInfo').textContent = `Spacing: ${spacing.toFixed(1)}"`;
            }
        } catch (e) {
            // Ignore parsing errors
        }
    }

    validateInputs() {
        try {
            const left = parseFloat(this.modal.querySelector('#leftBoundary').value);
            const right = parseFloat(this.modal.querySelector('#rightBoundary').value);
            const wallWidth = this.editor.wall.width;

            if (left < 0 || right > wallWidth) {
                alert('Boundaries must be within wall dimensions.');
                return [null, null];
            }

            if (left >= right) {
                alert('Left boundary must be less than right boundary.');
                return [null, null];
            }

            return [left, right];
        } catch (e) {
            alert('Please enter valid numbers for boundaries.');
            return [null, null];
        }
    }

    getYPosition() {
        try {
            const yPos = parseFloat(this.modal.querySelector('#yPosition').value);
            const wallHeight = this.editor.wall.height;

            if (yPos < 0 || yPos > wallHeight) {
                alert(`Y position must be between 0 and ${wallHeight} inches`);
                return null;
            }

            // Convert from floor distance to wall coordinates (from bottom)
            return wallHeight - yPos;
        } catch (e) {
            alert('Please enter a valid number for Y position');
            return null;
        }
    }

    async applySpacing() {
        const [left, right] = this.validateInputs();
        if (left === null) return;

        const yPosCenter = this.getYPosition();
        if (yPosCenter === null) return;

        if (this.selectedArtworks.length === 0) {
            alert('Please select at least one artwork.');
            return;
        }

        const totalWidth = this.selectedArtworks.reduce((sum, artwork) => sum + artwork.width, 0);
        const availableSpace = right - left;

        if (availableSpace < totalWidth) {
            alert('Not enough space for selected artworks.');
            return;
        }

        const spacing = (availableSpace - totalWidth) / (this.selectedArtworks.length + 1);

        if (spacing < this.MIN_SPACING) {
            if (!confirm(`Calculated spacing (${spacing.toFixed(1)}") is very small. Continue anyway?`)) {
                return;
            }
        }

        // Position artworks
        let currentX = left + spacing;

        for (const artwork of this.selectedArtworks) {
            // Calculate position in inches (wall coordinates)
            artwork.x_position = currentX;
            // Calculate Y position so center is at yPosCenter
            artwork.y_position = yPosCenter - (artwork.height / 2);

            // Update in database
            await this.editor.updateArtworkPosition(artwork.id, artwork.x_position, artwork.y_position);

            currentX += artwork.width + spacing;
        }

        // Refresh the display
        this.editor.renderArtworks();
        this.modalInstance.hide();
    }

    show() {
        // Reset selection
        this.selectedArtworks = [];
        this.modal.querySelector('#artworkList').selectedIndex = -1;
        this.updateDisplay();
        
        // Show modal
        this.modalInstance.show();
    }
}